src/imp/mbedtls.rs

```diff
@@ -0,0 +1,487 @@
+use self::mbedtls::pkcs12::{Pfx, Pkcs12Error};
+use self::mbedtls::hash::{Md, Type as MdType};
+use self::mbedtls::ssl::config::{Endpoint, Preset, Transport};
+use self::mbedtls::ssl::{Config, Context, Session, Version};
+use self::mbedtls::x509::certificate::List as CertList;
+use self::mbedtls::rng::{OsEntropy, CtrDrbg};
+use self::mbedtls::Error as TlsError;
+use self::mbedtls::Result as TlsResult;
+
+use std::error;
+use std::fmt;
+use std::io::{self, Read};
+use std::fs;
+
+use {Protocol, TlsAcceptorBuilder, TlsConnectorBuilder};
+
+fn load_ca_certs(dir: &str) -> TlsResult<Vec<::Certificate>> {
+    let paths = fs::read_dir(dir).map_err(|_| TlsError::X509FileIoError)?;
+
+    let mut certs = Vec::new();
+
+    for path in paths {
+
+        if let Ok(mut file) = fs::File::open(path.unwrap().path()) {
+            let mut contents = Vec::new();
+            if let Ok(_) = file.read_to_end(&mut contents) {
+                contents.push(0); // needs NULL terminator
+                if let Ok(cert) = ::Certificate::from_pem(&contents) {
+                    certs.push(cert);
+                }
+            }
+        }
+    }
+
+    Ok(certs)
+}
+
+#[derive(Debug)]
+pub enum Error {
+    Normal(TlsError),
+    Pkcs12(Pkcs12Error),
+    Custom(String),
+}
+
+impl From<TlsError> for Error {
+    fn from(err: TlsError) -> Error {
+        Error::Normal(err)
+    }
+}
+
+impl From<Pkcs12Error> for Error {
+    fn from(err: Pkcs12Error) -> Error {
+        Error::Pkcs12(err)
+    }
+}
+
+impl<S> From<TlsError> for HandshakeError<S> {
+    fn from(e: TlsError) -> HandshakeError<S> {
+        HandshakeError::Failure(e.into())
+    }
+}
+
+impl error::Error for Error {
+    fn description(&self) -> &str {
+        match *self {
+            Error::Normal(ref e) => error::Error::description(e),
+            Error::Pkcs12(ref e) => error::Error::description(e),
+            Error::Custom(ref e) => &e,
+        }
+    }
+
+    fn cause(&self) -> Option<&error::Error> {
+        match *self {
+            Error::Normal(ref e) => error::Error::cause(e),
+            Error::Pkcs12(ref e) => error::Error::cause(e),
+            Error::Custom(ref _e) => None,
+        }
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Error::Normal(ref e) => fmt::Display::fmt(e, fmt),
+            Error::Pkcs12(ref e) => fmt::Display::fmt(e, fmt),
+            Error::Custom(ref e) => fmt::Display::fmt(e, fmt),
+        }
+    }
+}
+
+fn map_version(protocol: Option<Protocol>) -> Option<Version> {
+    if let Some(protocol) = protocol {
+        match protocol {
+            Protocol::Sslv3 => Some(Version::Ssl3),
+            Protocol::Tlsv10 => Some(Version::Tls1_0),
+            Protocol::Tlsv11 => Some(Version::Tls1_1),
+            Protocol::Tlsv12 => Some(Version::Tls1_2),
+            _ => None
+        }
+    } else {
+        None
+    }
+
+}
+
+pub struct Identity(Pfx);
```

Comment:
Can this be refactored to be agnostic to PKCS12 or PKCS8? This is in the context of working to add PKCS8 support to the rest of native_tls's components

Reply:
Easily done, but until that support is landed in `TlsConnectorBuilder`/`TlsAcceptorBuilder` there is nothing to be done in this code since there is no way to create such an identity using the current public crate API.